### ISSUE TITLE
Add an option to force capitalizing the subject line

### ISSUE THREAD

**Issue by novadev94** [jorisroovers/gitlint#270]
This convention is well-known in the community and deemed a good practice in several articles related to Git commit messages, ex. https://cbea.ms/git-commit/#capitalize.

However, it's currently not possible to get it configured, due to Python stock regex module [`re`](https://docs.python.org/3/library/re.html) limitations - we can't have a regex pattern that matches a Unicode uppercase character.

There're 2 possible approaches to this:

* Implement this as a rule and _maybe_ a default one.

* Switch to [`regex`](https://pypi.org/project/regex/) module, which has richer regex support with full backward compatibility with the `re` module.

**Comment by sigmavirus24** [comment]
I don't think this must be done with regular expressions but it would be good to add even if it's off by default

**Comment by Bearwolves** [comment]
@novadev94 : Why is that not possible?

We are using following regex to check for uppercase character in the <description> from the conventional commit perspective [https://www.conventionalcommits.org/en/v1.0.0/](https://www.conventionalcommits.org/en/v1.0.0/) . I think that's what you are calling <subject>. 

```conf
[general]
...
contrib=contrib-title-conventional-commits

[title-match-regex]
# python-style regex that the commit-msg title must match
# Note that the regex can contradict with other rules if not used correctly
# (e.g. title-must-not-contain-word).
regex=^[a-z]+[(a-z)]*[!]*: [A-Z]+.*

[ignore-by-title]
# Ignore certain rules for commits of which the title matches a regex
# E.g. Match commit titles that start with "Release"
regex=^((?:Bump version from (.*) to)|(?:Bump version to)|(?:CHANGELOG\.md))(.*)

# Ignore certain rules, you can reference them by their id or by their full name
# Use 'all' to ignore all rules
ignore=contrib-title-conventional-commits, title-match-regex, body-min-length
```

What do I miss? :(

One thing to keep in mind, this new rule must not conflict with the contrib rules i.e. the conventional commits.

**Comment by novadev94** [comment]
@Bearwolves Yes, I should've called them "title" instead of "subject line" to match `gitlint` convention.
That conventional commit regex pattern has the same mentioned problem where `[A-Z]` only matches capital English letters, and doesn't match `Ö` for example.
That's a non-problem if we force commit messages to be English only, but imo that'd be too restrictive.

**Comment by Bearwolves** [comment]
@novadev94 : Got your point. Supporting other language/character shall be allowed, agreed!
Thanks for detailing!

**Comment by jorisroovers** [comment]
We can add this as a [contrib rule](https://jorisroovers.com/gitlint/contrib_rules/).

I’ve quickly implemented this a user-defined rule. Can you please verify this works as expected?

```python
# filename: captilized_title.py
from gitlint.rules import LineRule, RuleViolation, CommitMessageTitle

class CapitalizedTitle(LineRule):
    name = "capitalized-title"
    id = "UL1"
    target = CommitMessageTitle

    def validate(self, title, _commit):
        if (len(title) > 0 and not title[0].isupper()):
            return [ RuleViolation(self.id, "Title not capitalized", title) ]

        return []
```

Use this `.gitlint` config:

```python
[general]
extra-path=my/path/captilized_title.py
```

This would need unit tests and docs. Happy to accept a PR if you’re up for it: [https://jorisroovers.com/gitlint/contributing/#contrib-rules](https://jorisroovers.com/gitlint/contributing/#contrib-rules)

**Comment by jorisroovers** [comment]
Bump @novadev94 :-)
